sql/mysqld.cc

```diff
@@ -3195,11 +3210,16 @@ bool one_thread_per_connection_end(THD *thd, bool block_pthread)
   PSI_THREAD_CALL(delete_current_thread)();
 #endif
 
-  // Decrement connection count while holding LOCK_thread_count
+  mysql_mutex_lock(&LOCK_thread_count);
+  if (!global_thread_count.load(std::memory_order_relaxed)) {
+    mysql_cond_broadcast(&COND_thread_count);
+  }
+  dec_connection_count_locked();
+
   if (block_pthread)
-    block_pthread= block_until_new_connection();
+    block_pthread= block_until_new_connection_halflock();
```

Comment:
this one is tricky. The release of the lock happens midway within that function (for efficiency reasons ) and not at the very end.